### ISSUE TITLE
Proposal: in-memory secret store for dynamic secrets

### ISSUE THREAD

**Issue by jlongtine** [dagger/dagger#3442]
I need the ability to create new Secrets in the v0.3 API for `dagger-classic`. 

I suggest this change (or similar) to the API:

```graphql
query {
  secret { // make the id: SecretID optional
    new(contents: String!) Secret!
  }
}
```

Another possibility:

```graphql
query {
  secret(contents: String) Secret! // make the id: SecretID optional
}
```

**Comment by jlongtine** [comment]
@vito mentioned https://github.com/dagger/dagger/pull/3124

**Comment by samalba** [comment]
Discussed this briefly with @aluzzardi today to gather context and unblock @jlongtine.

Looks like there are two options:
- Option 1: (long-term), bigger effort: some sort of secret extension API to plug things like Vault, etc... as I understand that the ability to add secrets from the API (like in 0.2) is not gonna happen (for valid reasons). Also this API needs to be designed, reviewed and implemented. We know what it takes.
- Option 2: (short-term): have a private API (under engine, hidden from the Go SDK) to allow for creating secrets. This will be intended to disappear (risk to leak, and to stay forever).

I guess option 2 is more viable to unblock dagger-cue, depending on the effort for option 1. If we go option 2, we should find a way to derisk, for instance hiding the call behind a feature flag (eg. if this weird env var isn't set, the code path is disabled).

Would it be possible for @sipsma or @vito to own this one? Note that I am not hinting any option here, the priority is to remove the blocker. There might be an option 3 that I don't see.

**Comment by sipsma** [comment]
@jlongtine how absurd and awful of an idea is this: you currently can load secrets from env vars, so whenever you need to load a secret in dagger-cue you export an env var and then use that to load the secret in? Would need to be careful if `dagger-cue` ever forks off child processes of course. But could be a path to avoiding any changes to the API if it's plausible.

**Comment by shykes** [comment]
Maybe we should revisit the “no direct creation of secrets” rule, in case the circumstances have changed?

The initial motivation for “no direct creation” was to:

1. avoid leaking plaintext secrets into the “id” (serialized recipe). Because IDs are typically understood to be safe to copy paste, share etc.
2. avoid sending plaintext secrets on the wire

One major change is that we plan on making the engine remote; and our understanding of IDs and their role has evolved. So maybe it’s OK to create secrets directly after all?

TLDR my first guess would be either option 2 (hidden api call) or “option 2+” (regular api call). Skeptical about option 1 because it means rushing a pretty big design discussion for a short term problem.

**Comment by aluzzardi** [comment]
3. Bunch of edge cases we run into. The model of API-based creation broke down in a few places. For instance, when added in an extension it only works on un-cached runs (cached runs return a SecretID without the extension having run AddSecret, resulting in a broken secret)
4. To add to the above -- edge cases are caused by "AddSecret" being very special. It's the only call that alters the internal API state (whereas all other calls are idempotent, multi tenant, etc -- AddSecret is not)

I'd prefer we don't rush into neither Option 1 (needs design) nor Option 2+ (ripple effects, not easily reversible). @vito is going to post a summary of Option 3 (do nothing in core, work around with the existing API)

**Comment by vito** [comment]
You can actually create in-memory secrets already in a kind of round-about way:

```go
client.Directory().WithNewFile("sekrit", dagger.DirectoryWithNewFileOpts{
  Contents: "im always angry",
}).File("sekrit").Secret()
```

Caveat: this will result in a `SecretID` that embeds the secret value in plaintext (albeit obfuscated below multiple layers of base64 :sparkles: security :sparkles:). But maybe that's enough for now? As long a the `SecretID` itself isn't leaked this should still accomplish the goal of keeping its value out of the container layers.

This feels roughly the same level of risk as option 2 or introducing a dedicated "create secret" API, but the benefit is it already exists.

If you want to avoid leaking values into the `SecretID` you could instead leak them to the host filesystem and clean them up after:

```go
os.WriteFile("/sekrits/bruce", []byte("im always angry"), 0600)
defer os.Remove("/sekrits/bruce")

client.Host().Directory("/sekrits").File("bruce").Secret()

// ... do API calls ...
```

For extra pizazz you could _try_ using a fifo/pipe instead but we're skeptical that'll actually work. :P

Alternatively you could use env vars, but that's pretty sketchy too.

Between these two my preference would be the first one, since it's essentially what this issue proposes but feels like a workaround (which really it is).

**Comment by shykes** [comment]
If I understand correctly, Option 3 would be basically "all userland", no new API call either private or public? If so, that would be an ideal short term solution :)

Thanks for the additional details.

**Comment by jlongtine** [comment]
> @jlongtine how absurd and awful of an idea is this: you currently can load secrets from env vars, so whenever you need to load a secret in dagger-cue you export an env var and then use that to load the secret in? Would need to be careful if `dagger-cue` ever forks off child processes of course. But could be a path to avoiding any changes to the API if it's plausible.

@sipsma I've heard worse ideas in my time. :)

I think I'm going to start with this... see if works out, then move onto one of @vito's suggestions if it doesn't for whatever reason. 

I appreciate you all digging into this with/for me!

**Comment by jlongtine** [comment]
An update: Using environment variables works! I've got that all up and running, and it unblocked me on a bunch of stuff. Thanks for all the help, everyone!

**Comment by jlongtine** [comment]
I had a good discussion today with @sipsma and @vito about this. 

Our proposal as a first pass (with subsequent designs for extensions, etc) is to add a GraphQL API to allow the creation of new Secrets. This will add those secrets to the [`secret.Store`](https://github.com/dagger/dagger/blob/main/secret/store.go#LL17C8-L17C8), and allow for retrieval by Buildkit.

Everything feels pretty solid on implementation, but we still have some open questions around the API. 

### API Design 

I had previously suggested one or the other of these:

```graphqls
extend type Query {
  "Load a secret from its ID"
  secret(id: SecretID): Secret!
}

type Secret {
  "The identifier for this secret"
  id: SecretID!

  "The value of this secret"
  plaintext: String!

  "Create a new secret"
  new(plaintext: String!) Secret!
}
```

```graphqls
extend type Query {
  "Load a secret from its ID"
  secret(id: SecretID, plaintext: String): Secret!
}
```

With the following behavior:
```
if id == nil && plaintext == nil {
  return err
} else if id != nil && plaintext == nil {
  return the secret with id = id
} else if id != nil && plaintext != nil {
  create a new secret with this id and plaintext, unless a secret with this id already exists (and return an error in this case)
} else if id == nil && plaintext != nil {
  create a new secret with this plaintext, setting the id to sha256(plaintext)
}
```

@sipsma Also suggested this as a possibility:

```graphqls
extend type Query {
  withSecret(id: String!, plaintext: String!): Query!
}
```

which would allow queries like:

```graphql
query {
  withSecret(id: "mysecret", plaintext: "supersecretsecret") {
    container {
      from(address: "alpine") {
        withSecretVariable(name: "MYSECRET", secret: "mysecret") {
          withExec(args: ["printenv", "MYSECRET"]) {
            stdout
          }
        }
      }
    }
  }
}
```

@shykes, @aluzzardi Do you like either of these, or have any other ideas for this API?

**Comment by vito** [comment]
For completeness, here's another API we talked about, drawing a parallel to the `cacheVolume` constructor:

```graphql
extend type Query {
  inMemorySecret(plaintext: String!, name: String): Secret!
}
```

The `name` param is optional and defaults to a sha256 of the plaintext value. If specified, it can be used to prevent credential rotation from busting Buildkit caches.

There may be a better name than `inMemorySecret` but the general idea is that this API mutates the in-memory `session.Attachable` to register the secret and returns a `Secret` whose `SecretID` references its name. We also considered `sessionSecret` but avoided it since it forces us to decide whether the term "session" is official.

Note that this means the existing `Query { secret }` continues to have a required `id` param, whereas the `Secret { new }` proposal above implies that we make it optional (which also implies that there's an 'empty secret' representation).

**Comment by sipsma** [comment]
There's some big pros+cons to the approach of `withSecret(id: String!, plaintext: String!): Query!`.

In code it would ideally look like:
```go
c, err := dagger.Connect(ctx)
if err != nil {
  panic(err)
}

secretName := "mysecret"
c = c.WithSecret(secretName, "supersecretsecret")

ctr := c.Container().WithSecretVariable("SECRET", secretName)
// rest of your actual pipelines
```

Basically, the secret would always be provided as part of the queries you submit. In our codegen clients this would not be much overhead since it can be setup with a single line and then apply to all subsequent queries if you want.

Pros:
1. Secrets are only available for the duration of a single pipeline. Minimal scope, very explicit control, reflects the implementation where they exist only ephemerally in memory
2. Avoids questions about whether loading a secret should be a mutation or not; it just uses the same `with*` pattern we have elsewhere in our API
3. Relatively convenient to use on the codegen'd client level

Cons:
1. Painful to use on the raw graphql level since you have to explicitly wrap all your queries in `withSecret` when you want them to have access to those secrets.
   * Might be ways of alleviating this but goes off on other tangents
2. I don't know yet if our codegen clients handle the case of returning `Query`, if not they might need updates and thus increase implementation overhead

---

Overall, I like the pros of this option a lot, but I'm totally on board w/ other options too if there's concerns with this one. Any of them work for me 👍

**Comment by sipsma** [comment]
I keep thinking about this... I think our decision here is important not just because secret support is important but also because it might end up being the first time we directly expose session state in our actual API, which is worth doing consciously. 

To summarize, I think there's essentially two possible paths coming out of our decision here:
1. **Fully embrace session state** - this would be the implication of `secret` or `withInMemorySecret`.
   * Basically, this would be the first time we explicitly expose in our API the fact that there is mutable state associated with the session.
   * Previously there was some kind of implicit/hidden session state like resolution of tag->sha and syncing of local dirs, but we weren't sure if that's what we actually wanted in the long term (we just sort of inherited it from buildkit rather than designing it)
2. **Try to minimize session related state** - this would be the route of `withSecret` (or maybe other variations on that idea)
   * Secrets would just be scoped to individual queries; there's no state specific to the session that gets explicitly shared across queries

When we've discussed this in the past there's often been hesitance to go down the path of session statefulness since it introduces complexity, but on the other hand I guess it could just be inevitable anyways (services come to mind). I guess if we do go down the stateful path maybe we should just fully embrace it and make it an actual standard graphql API with mutations and such?

**Comment by aluzzardi** [comment]
Not sure what the desired end state is going to look like.

Apologies for not having read carefully through all the comments, but I see 2 approaches mentioned:

- A dedicated *new secret* endpoint (`inMemorySecret(plaintext: String!, name: String): Secret!`, `new(plaintext: String!) Secret!`)
- Carrying secrets on each request (session) (`c = c.WithSecret(secretName, "supersecretsecret")`)

IIRC last time we considered this, there were pitfalls on both approaches:

- Dedicated new secret endpoint breaks isolation
  - Generally speaking, every API query is currently self contained: order doesn't matter, everything is fully lazy, etc
  - The "new secret" endpoint would break this model
  - At the time, the problem was extensions (e.g. an extension call is cached, the secret ID is returned but the extension did not *add secret*)
  - Problem still exists, we just don't have extensions anymore. It does apply to dagger in dagger calls though, and will be relevant at some point
- Session
  - Fixes all the problems mentioned above
  - ... but IMHO it defeats the point of secrets a little bit
    - Might as well just pass plain text secrets around rather than `SecretID`, the end result wouldn't be so different

At this point I'm wondering what we want from secrets longer term. I'm worried that we might adopt an API that will make it harder to switch over to the "end goal"

**Comment by sipsma** [comment]
> Might as well just pass plain text secrets around rather than SecretID, the end result wouldn't be so different

The major difference is that the ID of objects wouldn't include the plaintext (what we want to avoid), it would only include the name of the secret. The query would have the plaintext (same as other options) but that only ever lives in memory.

It also could be expanded in the future to accept secrets that aren't initialized from plaintext; I think the same pattern should still be usable.

That being said, overall atm I'm actually leaning more towards inMemorySecret as I think that session statefulness is probably inevitable and we should just embrace it fully. Still thinking about it though.

**Comment by shykes** [comment]
Shouldn’t we be talking about the “pluggable secret stores” idea that we keep discussing as a future possibility? If we’re going to introduce a major new API design for secrets, it seems like a good time to explore that option. I know @aluzzardi that’s been on your mind.

I’m picturing a design (strawman) that leverages session state and attachables, without exposing them directly in the API. For example: l
  - start with cache volume style: no IDs, just names that map directly to buildkit secret id (plus maybe mangling)
  - on secret lookup by buildkit, session helper passes lookup to a *secret store*
  - builtin secret store just looks up the host environment. in an interactive tty it can prompt you interactively
  - secret store can be customized by the operator with a simple CLI hook, that can be set in `~/.config/dagger` or something. For example in GHA you would configure the CLI to look up gha secrets instead.
 
TLDR:
 - leverage session state yes
 - expose it in API no
 - there is a dimension missing here which is separation of concerns between pipeline dev & ops

**Comment by vito** [comment]
@shykes We actually started the discussion somewhere along those lines but figured the proposals so far (I'll call it 'static secrets') would be a faster way to deliver incremental value and unblock the Cue SDK.

Concretely, we discussed something like this (I'll call it 'dynamic secrets'):

```go
client := dagger.Connect(ctx, dagger.WithSecretProvider(func(ctx context.Context, secret string) ([]byte, error) {
  // fetch the secret somehow (either native library, or run an extension, whatever)
  return []byte("sekrit"), nil
}))
```

Personally I like this because it's super simple and avoids having to figure out Dagger's relationship with various secret providers, or whether they're extensions, etc. etc.; it's just a function. The architectural challenge is this requires a new call path from the engine to the client SDK.

On the other hand, we're pretty certain static secrets will be easy to implement and will immediately unblock the Cue SDK. It's also newly unblocked by the introduction of the session helper, which feels like a nice return on investment, but you could also read that as unnecessary bias. :wink: 

All that being said, I agree that we want dynamic secrets, so it would be nice if we can avoid painting ourselves into a corner just because we want to ship the Cue SDK sooner.

**Do we need to decide between "static OR dynamic secrets" or do we actually need "static AND dynamic secrets"?**

If it's "AND" I'd argue we continue with static secrets but make any adjustments necessary to be forward-compatible. i.e. if we think the advent of dynamic secrets deprecates `SecretID`, we should try to avoid it for static secrets too.

(I'm not sure if we can actually remove `SecretID` though - we also use it for e.g. `File { secret }` and `Host { envVariable { secret } }`.)

**Are there use cases served by static secrets that would be awkward if we only had dynamic secrets?**

I think so. Simple scripts might just take secrets from env vars and it'll be awkward to force that into a callback path. There are also cases where the client is constructed early in the lifecycle of a program at a point in time where it doesn't know what secrets will be needed in the future.

For example, [Bass constructs the Dagger client once on boot](https://github.com/vito/bass/blob/57110681c51d0aad536aae1cc7f74b11d316ce06/pkg/runtimes/dagger.go#L48) - it'll be awkward to use dynamic secrets here since the secrets don't exist until you _use_ the runtime. I could move the Dagger client construction to each runtime call, but that might worsen performance (assuming `dagger.Connect` does engine session setup/etc.)

**Comment by vito** [comment]
> For example, [Bass constructs the Dagger client once on boot](https://github.com/vito/bass/blob/57110681c51d0aad536aae1cc7f74b11d316ce06/pkg/runtimes/dagger.go#L48) - it'll be awkward to use dynamic secrets here since the secrets don't exist until you _use_ the runtime. I could move the Dagger client construction to each runtime call, but that might worsen performance (assuming `dagger.Connect` does engine session setup/etc.)

Also just realized this is actually an argument in favor of `withSecret` (https://github.com/dagger/dagger/issues/3442#issuecomment-1341787281) since I wouldn't need to carefully ensure secret names don't collide in shared session state.

**Comment by shykes** [comment]
Here's my proposal: https://github.com/dagger/dagger/pull/4193

**Comment by dolanor** [comment]
I don't have a lot of knowledge on BuildKit secret management. @jlongtine helped understand better but I feel I'm still not completely sure how it works.
I spent some time today in creating some diagram to share and be able to have feedback on my comprehension.

I had a first diagram that I presented to @jlongtine again.
(somehow, the mermaid interpreter from GitHub can't parse the code, so I put it in the annex)

![Capture d’écran 2022-12-14 à 18 57 04](https://user-images.githubusercontent.com/928722/207671700-4fec2afe-d42c-401f-beec-2cf12a3f4866.png)

So I thought the `dagger-session` would completely run in the Docker container, with the `graphql server` as well. But @jlongtine told me that the `graphql server` runs in the SDK (from the provisionning logic).

So the connection between the SDK and the Dagger Engine is not GraphQL based, but more gRPC for BuildKit Gateway?
![Capture d’écran 2022-12-14 à 20 06 37](https://user-images.githubusercontent.com/928722/207690353-339140ef-2da8-49a4-b341-11badd0515e9.png)

Then, the graph would look like that:
![Capture d’écran 2022-12-14 à 20 06 37](https://user-images.githubusercontent.com/928722/207690353-339140ef-2da8-49a4-b341-11badd0515e9.png)

So my questions are:
- What is right? What is wrong in those diagram?
- Is the Dagger Session in the SDK client, or in a container? (there was a `dagger-engine-session` container back in the old days, like, last week)
- Related: Is the session attachable client or server side?
- Are the Secrets fetched from the host or from buildkit
- how is the secret provider communicate with the host?

### Annex: mermaid code

#### 1st Graph: GraphQL Server in the Dagger Engine
<details>

```
graph LR
    subgraph Host
    subgraph UserCode[User Code]
        
        code --> nsfe(["core.NewSecretFromHostEnv('MY_SECRET')"])
    end

    

    UserCode --> SDK

    subgraph SDK
        gqlbuilder[GraphQL query builder]

        gqlbuilder -- build query --> hvs

        hvs[ HostVariable  secret ] ~~~ spc
        spc[Secrets Provider Connection] --> pt
        pt(["secret.PlainText()"]) --> ge(["os.GetEnv('MY_SECRET')"])
    end

    end

    SDK -.- provision -.-> DaggerEngine
    SDK -- GraphQL --> DaggerEngine
    subgraph Docker
        subgraph DaggerEngine
            direction LR

            subgraph GraphQLServer

                gs[GraphQL Server] -- GraphQL to LLB --> LLB[LLB DAG]
                LLB --- as(["llb.AddSecret( 'MY_SECRET' )"])
                
                as ~~~ sa

                sa[Session Attachable] ---> sp
                sp[Secrets Provider] --> getSecret(["GetSecret('MY_SECRET')"])
            end

            GraphQLServer -- query gRPC --> BuildKit

            subgraph BuildKit

                bk[BuildKit]

            end

            GraphQLServer -- session connection --- BuildKit
            BuildKit -- inject secrets from\n provider as file mounts --> Containers

            subgraph Containers

                c1[Container 1] --> c2[Container 2]
            end

            
        end

        
    end
    SDK == secrets provider session attachable connection === DaggerEngine
```

</details>

#### 2nd Graph: GraphQL server in the SDK/Client

<details>

```
graph TD
    subgraph Host
        direction LR

        subgraph UserCode[User Code]
            direction TB

            code --> wev(["WithEnvironmentVariable('MY_SECRET')"])
        end

        

        wev --> gqlbuilder

        subgraph SDK
            direction TB

            gqlbuilder[GraphQL query builder]

            gqlbuilder -- build query --> gs

            
            subgraph provisioner

            end

            subgraph GraphQLServer

                gs[GraphQL Request] -- GraphQL to LLB --> LLBDAG{{LLB DAG}}
                
                

            end

            subgraph DaggerSession
                bkgwClient[BuildKit Gateway client]
                sa[Session Attachable] --- sp
                sp[Secrets Provider] --> getSecret(["GetSecret('MY_SECRET')"])
                getSecret --> pt
            pt(["secretIDPayload{FromHostEnv: 'MY_SECRET'}.PlainText()"]) --> ge(["os.GetEnv('MY_SECRET')"])

            end

            
        end
    end
    

    bkSession -- "session connection (gRPC)" --- sa
    
    provisioner -- provision --> DaggerEngine

    
    subgraph Docker
        subgraph DaggerEngine
            direction LR



            

            subgraph BuildKit
                bkSession[Session]
                solver

                bkgwClient -- "query (gRPC)" --> bkSession
                bkSession --> llbDAG{{"LLB DAG"}}
                LLBDAG --> bkgwClient
                llbDAG --> as(["llb.AddSecret( 'MY_SECRET' )"])
                as -- GetSecret --> sp
            end

            ge ---> solver
            solver -- inject secrets from\n provider as file mounts --> Containers

            subgraph Containers

                c1[Container 1] --> c2[Container 2]
            end

            
        end

        
    end
```

</details>

**Comment by shykes** [comment]
Note, I gave this design proposal a nickname ("in-memory secret store") to make it easier to distinguish from other design proposals for handling dynamic secrets. See #4426 (this issue is listed as "option B").

**Comment by samalba** [comment]
<img width="500" alt="Screenshot 2023-01-24 at 2 41 55 PM" src="https://user-images.githubusercontent.com/216487/214437355-1eeb47b9-e64c-4edf-9cf9-f8a5e8ec49bf.png">

My 2c, the original name made more sense to me. This one implies a specific secret provider implementation, which wasn't the intent when @jlongtine opened it.

**Comment by shykes** [comment]
> My 2c, the original name made more sense to me. This one implies a specific secret provider implementation, which wasn't the intent when @jlongtine opened it.

The problem is that we have several possible ways of "creating a new secret via the API", and we need a way to distinguish them. Right now we have "in-memory store" (store the value of the secrets in memory) + "helpers" (retrieve the value of secrets from helpers). What names did you have in mind?
